Public/Src/Utilities/Instrumentation/Tracing/Log.cs

```diff
@@ -108,13 +108,55 @@ public abstract partial class Logger
             Message = "Managed memory: {0} MB, private bytes: {1} MB, threads: {2}")]
         public abstract void Memory(LoggingContext context, long managedHeapMegabytes, long privateMemorySizeMegabytes, int threads);
 
+        /// <summary>
+        /// Logs an UnexpectedCondition both locally as a warning and a capped number of times per build session to telemetry.
+        /// Always call this wrapper function
+        /// </summary>
+        /// <param name="loggingContext"></param>
+        /// <param name="description"></param>
+        public void UnexpectedCondition(LoggingContext loggingContext, string description)
+        {
+            lock (s_currentSessionId)

```

Comment:
>lock (s_currentSessionId) [](start = 12, length = 25)

I think it's a potential race condition: 
-- thread#1 takes a lock on an instance of `s_currentSessionId`
-- thread#1 reaches line 124 (`s_currentSessionId` is assigned a new variable) 
-- thread#2 enters `UnexpectedCondition` and it will successfully acquire another lock (because `s_currentSessionId` is a new instance now)

The worst thing that might happen is that we'll send more events to the telemetry; though it will only happen if we attempt to log multiple events from multiple threads at roughly the same time, and `s_currentSessionId` is not initialized. In other words, I'm not sure whether it warrants a change.

Reply:
Good catch. We share the same LoggingContext for the entire session so we shouldn't hit it. Easy enough to guard against though by using a separate object for the lock.